src/Data/Map/Internal.purs

```diff
@@ -218,126 +246,124 @@ isEmpty _ = false
-    comp :: k -> k -> Ordering
-    comp = compare
-
-    go Leaf = Nothing
-    go (Two left k1 v1 right) = case comp k k1 of
-      EQ -> findMax left
-      GT -> Just $ fromMaybe { key: k1, value: v1 } $ go right
-      LT -> go left
-    go (Three left k1 v1 mid k2 v2 right) = case comp k k2 of
-      EQ -> findMax $ Two left k1 v1 mid
-      GT -> Just $ fromMaybe { key: k2, value: v2 } $ go right
-      LT -> go $ Two left k1 v1 mid
+  go = case _ of
+    Leaf -> Nothing
+    Node _ _ mk mv ml mr ->
+      case compare k mk of
+        LT -> go ml
+        GT ->
+          case go mr of
+            Nothing -> Just { key: mk, value: mv }
+            other -> other
+        EQ ->
+          findMax ml
 
 -- | Look up a value for the specified key, or the least one greater than it
 lookupGE :: forall k v. Ord k => k -> Map k v -> Maybe { key :: k, value :: v }
 lookupGE k = go
   where
-    comp :: k -> k -> Ordering
-    comp = compare
-
-    go Leaf = Nothing
-    go (Two left k1 v1 right) = case comp k k1 of
-      EQ -> Just { key: k1, value: v1 }
-      LT -> Just $ fromMaybe { key: k1, value: v1 } $ go left
-      GT -> go right
-    go (Three left k1 v1 mid k2 v2 right) = case comp k k1 of
-      EQ -> Just { key: k1, value: v1 }
-      LT -> Just $ fromMaybe { key: k1, value: v1 } $ go left
-      GT -> go $ Two mid k2 v2 right
+  go = case _ of
+    Leaf -> Nothing
+    Node _ _ mk mv ml mr ->
+      case compare k mk of
+        LT ->
+          case go ml of
+            Nothing -> Just { key: mk, value: mv }
+            other -> other
+        GT -> go mr
+        EQ -> Just { key: mk, value: mv }
 
 -- | Look up a value for the least key greater than the specified key
 lookupGT :: forall k v. Ord k => k -> Map k v -> Maybe { key :: k, value :: v }
 lookupGT k = go
   where
-    comp :: k -> k -> Ordering
-    comp = compare
-
-    go Leaf = Nothing
-    go (Two left k1 v1 right) = case comp k k1 of
-      EQ -> findMin right
-      LT -> Just $ fromMaybe { key: k1, value: v1 } $ go left
-      GT -> go right
-    go (Three left k1 v1 mid k2 v2 right) = case comp k k1 of
-      EQ -> findMin $ Two mid k2 v2 right
-      LT -> Just $ fromMaybe { key: k1, value: v1 } $ go left
-      GT -> go $ Two mid k2 v2 right
+  go = case _ of
+    Leaf -> Nothing
+    Node _ _ mk mv ml mr ->
+      case compare k mk of
+        LT ->
+          case go ml of
+            Nothing -> Just { key: mk, value: mv }
+            other -> other
+        GT -> go mr
+        EQ -> findMin mr
 
 -- | Returns the pair with the greatest key
 findMax :: forall k v. Map k v -> Maybe { key :: k, value :: v }
-findMax = go Nothing
-  where
-    go acc Leaf = acc
-    go _   (Two _ k1 v1 right) = go (Just { key: k1, value: v1 }) right
-    go _   (Three _ _ _ _ k2 v2 right) = go (Just { key: k2, value: v2 }) right
+findMax = case _ of
+  Leaf ->
+    Nothing
+  Node _ _ k v _ Leaf ->
+    Just { key: k, value: v }
+  Node _ _ _ _ _ r ->
+    findMax r
```

Comment:
In code above this (e.g. `eq`), there's no redundant pattern matching. And yet here, you don't write:
```
case _ of
  Leaf -> ..
  Node _ _ _ _ _ r -> case r of
    Leaf -> ...
    _ -> ...
```

Any particular reason?